src/ocl_dev_mgr.cpp

```diff
@@ -4,365 +4,409 @@
+            }
+        }
     }
-  }
 
-  return &(con_list.at(context_idx).kernels.at(idx).at(0));
+    return &(con_list.at(context_idx).kernels.at(idx).at(0));
 }
 
 cl::Kernel* ocl_dev_mgr::getKernelbyID(cl_uint context_idx, std::string const& prog_name, cl_ulong kernel_id)
 {
-  auto it_p = find(con_list.at(context_idx).prog_names.begin(), con_list.at(context_idx).prog_names.end(), prog_name);
-  if (it_p == con_list.at(context_idx).prog_names.end()) {
-    return nullptr;
-  }
+    auto it_p = find(con_list.at(context_idx).prog_names.begin(), con_list.at(context_idx).prog_names.end(), prog_name);
+    if (it_p == con_list.at(context_idx).prog_names.end()) {
+        return nullptr;
+    }
 
-  uint32_t idx = distance(con_list.at(context_idx).prog_names.begin(), it_p);
+    uint32_t idx = distance(con_list.at(context_idx).prog_names.begin(), it_p);
 
-  return &(con_list.at(context_idx).kernels.at(idx).at(kernel_id));
+    return &(con_list.at(context_idx).kernels.at(idx).at(kernel_id));
 }
 
 
 std::string ocl_dev_mgr::getDeviceType(cl_uint avail_device_idx)
 {
-  if (available_devices.at(avail_device_idx).type == CL_DEVICE_TYPE_CPU) {
-    return(type_cpu_str);
-  }
-  else if (available_devices.at(avail_device_idx).type == CL_DEVICE_TYPE_GPU) {
-    return(type_gpu_str);
-  }
-  else if (available_devices.at(avail_device_idx).type == CL_DEVICE_TYPE_ACCELERATOR) {
-    return(type_acc_str);
-  }
-  else {
-    return(type_other_str);
-  }
+    if (available_devices.at(avail_device_idx).type == CL_DEVICE_TYPE_CPU) {
+        return(type_cpu_str);
+    }
+    else if (available_devices.at(avail_device_idx).type == CL_DEVICE_TYPE_GPU) {
+        return(type_gpu_str);
+    }
+    else if (available_devices.at(avail_device_idx).type == CL_DEVICE_TYPE_ACCELERATOR) {
+        return(type_acc_str);
+    }
+    else {
+        return(type_other_str);
+    }
 }
 
+std::string ocl_dev_mgr::getDevicePCIeID(cl_uint avail_device_idx)
+{
+#define CL_DEVICE_PCI_BUS_ID_NV  0x4008
+#define CL_DEVICE_PCI_SLOT_ID_NV 0x4009
+#define CL_DEVICE_TOPOLOGY_AMD   0x4037
+typedef union
+{
+    struct { cl_uint type; cl_uint data[5]; } raw;
+    struct { cl_uint type; cl_char unused[17]; cl_char bus; cl_char device; cl_char function; } pcie;
+} cl_device_topology_amd;
+
+    cl_device_topology_amd amd_topo;
+	cl_int bus_id;
+	cl_int slot_id;
+	std::ostringstream tmp_stream;
+
+    std::size_t found = 0;
+    found = available_devices.at(avail_device_idx).vendor.find("NVIDIA");
+    if (found != std::string::npos) {
+        available_devices.at(avail_device_idx).device.getInfo(CL_DEVICE_PCI_BUS_ID_NV,&bus_id);
+	    available_devices.at(avail_device_idx).device.getInfo(CL_DEVICE_PCI_SLOT_ID_NV, &slot_id);
+
+	    cl_uint domain, bus, dev, func;
+	    domain = bus_id >> 8;
+	    bus = bus_id & 0xff;
+	    dev = slot_id >> 3;
+	    func = slot_id & 0x7;
+	    tmp_stream << domain << ":" << bus << ":" << slot_id;
```

Comment:
`dev` and `func` are not really used?

Reply:
Not really, as the NVML format is strange, I left it to remember how to shift the bits to get separate device and function ids, if needed.

Reply:
Maybe you could comment these lines out and add a corresponding remark? Right now, some compilers show warnings (and the warnings could confuse users).

Reply:
Right now I removed it from the offical version as I realized that we can just use the git history.